### PR TITLE
Adding a cleanup function

### PR DESCRIPTION
**Pull request by jmlarson1**
This cleans up the main `pounders` routine and puts some (but not all) of the statements before exiting into one location.

### CODE REVIEW

**Comment by jared321** [comment]
One thing that we can try to work toward if we continue evolving this is to improve maintenance, especially if we adopt this as a POptUS-wide design.  You presently have an explanation for each flag value
* in the `pounders.m` documentation,
* in the `cleanup_before_return.m` documentation, and
* in the code that returns the error message.

That's a lot of duplication and plenty of room to have bad documentation.  I presently believe that some of this arises from decoupling the error message, which POUNDerS emits under the hood, from the error flag, which is passed to the application layer `callpounders` for handling.

Also, if the errors document themselves well enough, does the documentation for the functions also need to document them?

File: pounders/m/clean_up_before_return.m [hunk 1]
```diff
@@ -0,0 +1,25 @@
+% This function is called to cleanup X and F, set the exit value, and display
+% reason for exiting
+
+% flag    [dbl] Termination criteria flag:
+%               = 0 normal termination because of grad,
+%               > 0 exceeded nfmax evals,   flag = norm of grad at final X
+%               = -1 if input was fatally incorrect (error message shown)
+%               = -2 model failure
+%               = -3 error from TRSP Solver or if a NaN was encountered
```
Comment (jared321) [inline pounders/m/clean_up_before_return.m]: This is different from the flag documentation in `pounders.m`.  Which is correct?

File: pounders/m/clean_up_before_return.m [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+%               = -4 error in TRSP Solver
+
+function [X, F, exit_flag] = clean_up_before_return(X, F, nf, exit_flag)
+
+    X = X(1:nf, :);
+    F = F(1:nf, :);
+
+    if exit_flag == -4
+        disp("An occured in minq. Exiting.");
```
Comment (jared321) [inline pounders/m/clean_up_before_return.m:10]: The original MINQ error message was more specific.  Aren't those details potentially useful to the user?

**Comment by jmlarson1** [comment]
@jared321 Thank you for pointing this out. I've removed the redundant documentation from the two `cleanup_before_return` files to resolve much of the issue. What do you think now?

File: pounders/py/clean_up_before_return.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+#               = -1 if input was fatally incorrect (error message shown)
+#               = -2 model failure
+#               = -3 error from TRSP Solver or if a NaN was encountered
+#               = -4 error in TRSP Solver
+
+
+def clean_up_before_return(X, F, nf, exit_flag):
+    X = X[: nf + 1]
+    F = F[: nf + 1]
+
```
Comment (jared321) [inline pounders/py/clean_up_before_return.py:8]: The python code below was cleaning up with `nf` and `nf+1`.  MATLAB was cleaning up consistently.  Is this forcing python to obey a single clean-up rule or a bug fix?

**Comment by jared321** [comment]
Certainly better.  I wonder if we can improve more though.

You have watered down some of the error messages.  For instance, with the current error handling scheme, will users be able to tell where in `pounders.m` a type -2 error was thrown?

**Comment by jared321** [comment]
My two cents, your routines aren't cleaning up.  The routines should be used solely for the purpose of setting outputs before return.

**Review by jared321 — APPROVED** [review]

File: pounders/m/clean_up_before_return.m [hunk 4]
```diff
@@ -0,0 +1,25 @@
+% This function is called to cleanup X and F, set the exit value, and display
+% reason for exiting
+
+% flag    [dbl] Termination criteria flag:
```
[the remark this answers is not in the document]
Comment (jmlarson1) [reply pounders/m/clean_up_before_return.m]: Yes, and your suggestion removed this issue.

File: pounders/m/clean_up_before_return.m (hunk 1 above)
Comment (jmlarson1) [reply pounders/m/clean_up_before_return.m]: As you pointed out, removing redundant documentation helped remove this issue.

File: pounders/py/clean_up_before_return.py (hunk 3 above)
Comment (jmlarson1) [reply pounders/py/clean_up_before_return.py:8]: There was a bug (or at least unexpected behavior) before! 

In `.py` we weren't returning the last entries of `X` and `F` when `F` had a NaN. This further highlights the need for such a single location for cleaning up.

File: pounders/m/clean_up_before_return.m (hunk 2 above)
Comment (jmlarson1) [reply pounders/m/clean_up_before_return.m]: Corrected!

**Comment by jmlarson1** [comment]
> My two cents, your routines aren't cleaning up. The routines should be used solely for the purpose of setting outputs before return.

I used "cleaning up" because it's removing unfilled rows of `X` and `F`. Is that not a "clean up" procedure?

**Comment by jared321** [comment]
I would normally associate clean-up with freeing up resources, etc. as part of a graceful failure.  Your routine is preparing `X`, `F` specifically so that `pounders.m` can return them to the calling code.  It makes your routine special purpose.  Not only should you be calling it in only a few certain circumstances (normal), but you should always assign the outputs to the same set of "special" variables.

**Pull request merged**